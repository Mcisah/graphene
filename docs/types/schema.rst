Schema
======

A Schema is created by supplying the root types of each type of operation, query, mutation and subscription.
A schema definition is then supplied to the validator and executor.

.. code:: js

    import { Schema } from "graphene-js";

    const schema = new Schema({
        query: MyRootQuery,
        mutation: MyRootMutation,
    })

Types
-----

There are some cases where the schema cannot access all of the types that we plan to have.
For example, when a field returns an ``Interface``, the schema doesn't know about any of the
implementations.

In this case, we need to use the ``types`` argument when creating the Schema.


.. code:: js

    const schema = new Schema({
        query: MyRootQuery,
        types=[SomeExtraType, ],
    })


Querying
--------

To query a schema, call the ``execute`` method on it.


.. code:: js

    await schema.execute('{ hello }')

