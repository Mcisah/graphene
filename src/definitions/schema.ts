/**
 * Copyright (c) 2017-present, Graphene.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import {
  GraphQLObjectType,
  ExecutionResult,
  GraphQLDirective,
  GraphQLSchema,
  GraphQLNamedType,
  graphql,
  printSchema
} from 'graphql';

import { getGraphQLType } from '../reflection';

// The provided configuration type when creating a new Schema.
export type SchemaConfig = {
  query: any;
  mutation?: any;
  subscription?: any;
  directives?: GraphQLDirective[];
  types?: any[];
};

type GraphQLSchemaConfig = {
  query: GraphQLObjectType;
  mutation?: GraphQLObjectType;
  subscription?: GraphQLObjectType;
  directives?: GraphQLDirective[];
  types?: GraphQLNamedType[];
};

export class Schema extends GraphQLSchema {
  constructor(config: SchemaConfig) {
    let schemaConfig: GraphQLSchemaConfig = {
      query: <GraphQLObjectType>getGraphQLType(config.query),
      directives: config.directives
    };
    if (config.mutation) {
      schemaConfig.mutation = <GraphQLObjectType>getGraphQLType(
        config.mutation
      );
    }
    if (config.subscription) {
      schemaConfig.subscription = <GraphQLObjectType>getGraphQLType(
        config.subscription
      );
    }
    if (config.types) {
      schemaConfig.types = config.types.map(
        type => <GraphQLObjectType>getGraphQLType(type)
      );
    }
    super(schemaConfig);
  }
  public execute(query: string, ...args: any[]): Promise<ExecutionResult> {
    return graphql(this, query, ...args);
  }
  public toString() {
    return printSchema(this);
  }
}
