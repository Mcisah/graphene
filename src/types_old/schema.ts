import {
  GraphQLSchema,
  GraphQLObjectType,
  graphql,
  printSchema
} from 'graphql';

import { getGraphQLType } from './base';
import { ObjectType } from './objecttype';

type SchemaOptions = {
  query: typeof ObjectType | GraphQLObjectType;
  mutation?: typeof ObjectType | GraphQLObjectType;
  subscription?: typeof ObjectType | GraphQLObjectType;
};

export class Schema extends GraphQLSchema {
  constructor(options: SchemaOptions) {
    super({
      ...options,
      query: <GraphQLObjectType>getGraphQLType(options.query),
      mutation: options.mutation
        ? <GraphQLObjectType>getGraphQLType(options.mutation)
        : undefined,
      subscription: options.subscription
        ? <GraphQLObjectType>getGraphQLType(options.subscription)
        : undefined
    });
  }
  execute = graphql.bind(null, this);
  toString() {
    return printSchema(this);
  }
}
