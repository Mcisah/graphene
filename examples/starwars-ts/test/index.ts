import schema, { Episode } from '../schema4';

describe('Interface', () => {
  test(`create Interface`, () => {
    expect(schema.toString()).toMatchSnapshot();
  });
  test('try query', async () => {
    var result = await schema.execute(`
{
    hero {
        id
        name
        friends {
            id
            name
        }
    }
}
`);
    expect(result).toMatchObject({
      data: {
        hero: {
          friends: [
            { id: '1000', name: 'Luke Skywalker' },
            { id: '1002', name: 'Han Solo' },
            { id: '1003', name: 'Leia Organa' }
          ],
          id: '2001',
          name: 'R2-D2'
        }
      }
    });
  });
  test('Episode enum', () => {
    expect(Episode.NEWHOPE).toBe(4);
    expect(Episode.EMPIRE).toBe(5);
    expect(Episode.JEDI).toBe(6);
  });
});
