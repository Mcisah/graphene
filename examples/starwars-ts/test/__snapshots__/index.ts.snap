// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interface create Interface 1`] = `
"\\"\\"\\"A character in the Star Wars Trilogy\\"\\"\\"
interface Character {
  \\"\\"\\"The id of the character.\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character.\\"\\"\\"
  name: String

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends: [Character]

  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn: [Episode]
}

\\"\\"\\"A mechanical creature in the Star Wars universe.\\"\\"\\"
type Droid implements Character {
  \\"\\"\\"The id of the character.\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character.\\"\\"\\"
  name: String

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends: [Character]

  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn: [Episode]

  \\"\\"\\"The primary function of the droid, or null if unknown..\\"\\"\\"
  primaryFunction: String
}

\\"\\"\\"The description\\"\\"\\"
enum Episode {
  \\"\\"\\"Released in 1977.\\"\\"\\"
  NEWHOPE

  \\"\\"\\"Released in 1980.\\"\\"\\"
  EMPIRE

  \\"\\"\\"Released in 1983.\\"\\"\\"
  JEDI
}

\\"\\"\\"A humanoid creature in the Star Wars universe.\\"\\"\\"
type Human implements Character {
  \\"\\"\\"The id of the character.\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character.\\"\\"\\"
  name: String

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends: [Character]

  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn: [Episode]

  \\"\\"\\"The home planet of the human, or null if unknown..\\"\\"\\"
  homePlanet: String
}

type Query {
  hero(episode: Episode): Character
  human(id: String): Human
  droid(id: String): Droid
}
"
`;
